File: phantom_native/zig/src/neurocore.zig
// neurocore.zig — Sovereign NeuroCore SIMD Kernels
//
// Production-grade native kernels for the Phantom-MESIE stack.
// Uses explicit SIMD intrinsics for vectorized spectral operations.
// C ABI exports for Python/CFFI binding.

const std = @import("std");
const math = std.math;

// ============================================================
// SovereignTensor — Cache-friendly f32 buffer with metadata
// ============================================================

pub const SpectralMeta = struct {
    resonance: f32 = 1.0,
    helix_turns: u32 = 8,
    quantized: bool = false,
    quant_scale: f32 = 1.0,
};

pub const SovereignTensor = struct {
    data: []f32,
    shape: []const usize,
    meta: SpectralMeta,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, shape: []const usize, meta: SpectralMeta) !SovereignTensor {
        var size: usize = 1;
        for (shape) |d| {
            size *= d;
        }
        const data = try allocator.alloc(f32, size);
        @memset(data, 0.0);
        return SovereignTensor{
            .data = data,
            .shape = shape,
            .meta = meta,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *SovereignTensor) void {
        self.allocator.free(self.data);
    }

    pub fn size(self: SovereignTensor) usize {
        return self.data.len;
    }
};

// ============================================================
// SIMD Vectorized Operations
// ============================================================

/// 8-wide SIMD vector addition with scalar tail handling.
/// Resonance product is applied to the result.
pub fn vector_add_simd(a: SovereignTensor, b: SovereignTensor, out: []f32) void {
    const n = @min(a.data.len, b.data.len);
    const resonance = a.meta.resonance * b.meta.resonance;

    var i: usize = 0;

    // 8-wide SIMD loop
    while (i + 8 <= n) : (i += 8) {
        const va: @Vector(8, f32) = a.data[i..][0..8].*;
        const vb: @Vector(8, f32) = b.data[i..][0..8].*;
        const vr = va + vb;
        const res_vec: @Vector(8, f32) = @splat(resonance);
        const scaled = vr * res_vec;
        out[i..][0..8].* = scaled;
    }

    // Scalar tail
    while (i < n) : (i += 1) {
        out[i] = (a.data[i] + b.data[i]) * resonance;
    }
}

/// 8-wide SIMD element-wise multiplication.
pub fn vector_mul_simd(a: SovereignTensor, b: SovereignTensor, out: []f32) void {
    const n = @min(a.data.len, b.data.len);

    var i: usize = 0;
    while (i + 8 <= n) : (i += 8) {
        const va: @Vector(8, f32) = a.data[i..][0..8].*;
        const vb: @Vector(8, f32) = b.data[i..][0..8].*;
        out[i..][0..8].* = va * vb;
    }
    while (i < n) : (i += 1) {
        out[i] = a.data[i] * b.data[i];
    }
}

/// Resonance-weighted matrix multiplication with 4-wide inner unrolling.
/// For fixed spectral shapes (e.g., 128x128, 64x64).
pub fn resonance_matmul(
    a: SovereignTensor,
    b: SovereignTensor,
    out: []f32,
    m: usize,
    k: usize,
    n: usize,
) void {
    const resonance = a.meta.resonance * b.meta.resonance;

    for (0..m) |i| {
        for (0..n) |j| {
            var acc: f32 = 0.0;
            var p: usize = 0;

            // 4-wide inner unroll
            while (p + 4 <= k) : (p += 4) {
                acc += a.data[i * k + p] * b.data[p * n + j] +
                    a.data[i * k + p + 1] * b.data[(p + 1) * n + j] +
                    a.data[i * k + p + 2] * b.data[(p + 2) * n + j] +
                    a.data[i * k + p + 3] * b.data[(p + 3) * n + j];
            }
            // Scalar tail
            while (p < k) : (p += 1) {
                acc += a.data[i * k + p] * b.data[p * n + j];
            }

            out[i * n + j] = acc * resonance;
        }
    }
}

// ============================================================
// Resonance Attention Kernel
// ============================================================

/// Native resonance attention: dot-product with exponential decay + softmax.
pub fn resonance_attention(
    q: []const f32,
    k: []const f32,
    v: []const f32,
    out: []f32,
    n: usize,
) void {
    const scale: f32 = 1.0 / @sqrt(@as(f32, @floatFromInt(n)));

    // Compute scores with resonance decay
    var scores: [512]f32 = undefined;
    var max_score: f32 = -math.inf(f32);

    for (0..n) |i| {
        const dot = q[i] * k[i] * scale;
        const resonance = @exp(-@abs(dot) * 0.5);
        scores[i] = dot * resonance;
        if (scores[i] > max_score) max_score = scores[i];
    }

    // Stable softmax
    var total: f32 = 0.0;
    for (0..n) |i| {
        scores[i] = @exp(scores[i] - max_score);
        total += scores[i];
    }
    if (total == 0.0) total = 1.0;

    // Weighted output
    for (0..n) |i| {
        out[i] = (scores[i] / total) * v[i];
    }
}

// ============================================================
// Helix Encoding
// ============================================================

/// Apply helix rotation encoding for spectral retrieval.
pub fn helix_encode(data: []const f32, out: []f32, n: usize, turns: u32) void {
    const n_f: f32 = @floatFromInt(if (n > 1) n - 1 else 1);
    const turns_f: f32 = @floatFromInt(turns);

    for (0..n) |i| {
        const i_f: f32 = @floatFromInt(i);
        const phase = 2.0 * math.pi * turns_f * (i_f / n_f);
        out[i] = data[i] * @cos(phase) + @sin(phase) * 0.1;
    }
}

// ============================================================
// INT8 Quantization for Edge Deployment
// ============================================================

/// Quantize f32 tensor to INT8 range [-127, 127].
pub fn quantize_int8(data: []const f32, out: []i8, n: usize) f32 {
    var max_abs: f32 = 0.0;
    for (0..n) |i| {
        const abs_val = @abs(data[i]);
        if (abs_val > max_abs) max_abs = abs_val;
    }
    if (max_abs == 0.0) max_abs = 1.0;

    for (0..n) |i| {
        const scaled = (data[i] / max_abs) * 127.0;
        out[i] = @intFromFloat(@min(@max(scaled, -127.0), 127.0));
    }

    return max_abs; // return scale for dequantization
}

// ============================================================
// C ABI Exports (for Python CFFI / ctypes binding)
// ============================================================

export fn phantom_vector_add(a: [*]const f32, b: [*]const f32, out: [*]f32, n: u32) void {
    const len = @as(usize, n);
    var i: usize = 0;
    while (i + 8 <= len) : (i += 8) {
        const va: @Vector(8, f32) = a[i..][0..8].*;
        const vb: @Vector(8, f32) = b[i..][0..8].*;
        out[i..][0..8].* = va + vb;
    }
    while (i < len) : (i += 1) {
        out[i] = a[i] + b[i];
    }
}

export fn phantom_resonance_matmul(
    a: [*]const f32,
    b: [*]const f32,
    out: [*]f32,
    m: u32,
    k: u32,
    n: u32,
    resonance: f32,
) void {
    const m_sz = @as(usize, m);
    const k_sz = @as(usize, k);
    const n_sz = @as(usize, n);

    for (0..m_sz) |i| {
        for (0..n_sz) |j| {
            var acc: f32 = 0.0;
            for (0..k_sz) |p| {
                acc += a[i * k_sz + p] * b[p * n_sz + j];
            }
            out[i * n_sz + j] = acc * resonance;
        }
    }
}

export fn phantom_helix_encode(data: [*]const f32, out: [*]f32, n: u32, turns: u32) void {
    const len = @as(usize, n);
    const n_f: f32 = @floatFromInt(if (len > 1) len - 1 else 1);
    const turns_f: f32 = @floatFromInt(turns);

    for (0..len) |i| {
        const i_f: f32 = @floatFromInt(i);
        const phase = 2.0 * math.pi * turns_f * (i_f / n_f);
        out[i] = data[i] * @cos(phase) + @sin(phase) * 0.1;
    }
}

// ============================================================
// Tests
// ============================================================

test "vector_add_simd basic" {
    const allocator = std.testing.allocator;
    var a_data = try allocator.alloc(f32, 16);
    defer allocator.free(a_data);
    var b_data = try allocator.alloc(f32, 16);
    defer allocator.free(b_data);
    var out = try allocator.alloc(f32, 16);
    defer allocator.free(out);

    for (0..16) |i| {
        a_data[i] = @floatFromInt(i);
        b_data[i] = 1.0;
    }

    const shape = &[_]usize{16};
    const a = SovereignTensor{ .data = a_data, .shape = shape, .meta = .{}, .allocator = allocator };
    const b = SovereignTensor{ .data = b_data, .shape = shape, .meta = .{}, .allocator = allocator };

    vector_add_simd(a, b, out);

    try std.testing.expectApproxEqAbs(out[0], 1.0, 0.001);
    try std.testing.expectApproxEqAbs(out[7], 8.0, 0.001);
    try std.testing.expectApproxEqAbs(out[15], 16.0, 0.001);
}

test "resonance_attention basic" {
    var q = [_]f32{ 1.0, 0.5, 0.3, 0.1 };
    var k_arr = [_]f32{ 0.5, 1.0, 0.2, 0.8 };
    var v = [_]f32{ 1.0, 2.0, 3.0, 4.0 };
    var out: [4]f32 = undefined;

    resonance_attention(&q, &k_arr, &v, &out, 4);

    // Output should be non-zero weighted values
    var total: f32 = 0.0;
    for (out[0..4]) |val| {
        total += val;
        try std.testing.expect(val >= 0.0);
    }
    try std.testing.expect(total > 0.0);
}

File: phantom_native/neurocore.py
"""SovereignNeuroCore — Native MESIE NeuroCore with resonance attention.

A self-contained neural processing unit combining:
- Helix-encoded weight initialization
- Resonance-weighted attention kernels
- TAURUS memory integration (working + long-term)
- SIMD-style vectorized forward pass

Zero heavy dependencies — uses only stdlib + SovereignTensor + TaurusMemory.
"""

from __future__ import annotations

import math
from typing import Any, Dict, List, Optional

from phantom_native.sovereign_tensor import SovereignTensor
from phantom_native.taurus import TaurusMemory


class SovereignNeuroCore:
    """Native MESIE NeuroCore — resonance + helix + TAURUS aware.

    Attributes:
        d_model: Internal representation dimension.
        n_heads: Number of attention heads.
        taurus: Native TAURUS memory instance.
        weights: Helix-encoded weight matrices (Q, K, V).
    """

    def __init__(self, config: Optional[Dict[str, Any]] = None):
        config = config or {}
        self.d_model = config.get("d_model", 128)
        self.n_heads = config.get("n_heads", 8)
        self.head_dim = self.d_model // self.n_heads
        self.taurus = TaurusMemory(
            capacity=config.get("memory_capacity", 32),
            decay_rate=config.get("decay_rate", 0.95),
        )
        self.weights = self._init_helix_weights()
        self._attention_maps: List[List[float]] = []

    def _init_helix_weights(self) -> Dict[str, List[float]]:
        """Initialize weights using helix encoding from MESIE primitives.

        Helix initialization provides structured frequency-aware starting
        points that align with spectral data characteristics.
        """
        d = self.d_model
        return {
            "query": [math.sin(i * 0.1) * 0.5 for i in range(d)],
            "key": [math.cos(i * 0.1) * 0.5 for i in range(d)],
            "value": [1.0 / math.sqrt(d) for _ in range(d)],
            "output": [math.sin(i * 0.05 + 0.3) * 0.3 for i in range(d)],
        }

    def _resonance_attention(
        self, q: List[float], k: List[float], v: List[float]
    ) -> List[float]:
        """Resonance-weighted attention kernel.

        Computes attention scores with exponential resonance decay,
        then applies softmax and weighted sum over values.

        Args:
            q: Query vector.
            k: Key vector.
            v: Value vector.

        Returns:
            Attention-weighted output vector.
        """
        n = len(q)
        scores: List[float] = []

        # Compute resonance-weighted attention scores
        scale = 1.0 / math.sqrt(n) if n > 0 else 1.0
        for i in range(n):
            dot = q[i] * k[i] * scale
            resonance = math.exp(-abs(dot) * 0.5)  # resonance decay
            scores.append(dot * resonance)

        # Softmax (numerically stable)
        max_s = max(scores) if scores else 0.0
        exp_s = [math.exp(s - max_s) for s in scores]
        total = sum(exp_s) or 1.0
        attn_weights = [e / total for e in exp_s]

        # Store attention map for interpretability
        self._attention_maps.append(attn_weights[:])

        # Weighted sum over values
        output = [0.0] * n
        for i in range(n):
            output[i] = attn_weights[i] * v[i]

        return output

    def forward(self, tensor: SovereignTensor) -> SovereignTensor:
        """Full forward pass with resonance attention, helix projection, and TAURUS.

        Pipeline:
        1. Project input to Q, K, V via helix weights
        2. Apply resonance attention kernel
        3. Output projection
        4. Store result in TAURUS working memory
        5. Fuse with top-k context from memory

        Args:
            tensor: Input SovereignTensor (any 1D shape).

        Returns:
            Processed SovereignTensor with attention-weighted embedding.
        """
        self._attention_maps.clear()
        d = self.d_model
        n = len(tensor.data)

        # Project to QKV (cyclic indexing for arbitrary input sizes)
        q = [
            tensor.data[i % n] * self.weights["query"][i % d] for i in range(d)
        ]
        k = [
            tensor.data[i % n] * self.weights["key"][i % d] for i in range(d)
        ]
        v = [
            tensor.data[i % n] * self.weights["value"][i % d] for i in range(d)
        ]

        # Multi-head resonance attention
        head_outputs: List[List[float]] = []
        for h in range(self.n_heads):
            start = h * self.head_dim
            end = start + self.head_dim
            head_q = q[start:end]
            head_k = k[start:end]
            head_v = v[start:end]
            head_out = self._resonance_attention(head_q, head_k, head_v)
            head_outputs.append(head_out)

        # Concatenate heads
        concat = []
        for head_out in head_outputs:
            concat.extend(head_out)

        # Output projection
        output_data = [
            concat[i] * self.weights["output"][i % d] for i in range(len(concat))
        ]

        # Pad or truncate to match original input shape
        if len(output_data) < n:
            output_data.extend([0.0] * (n - len(output_data)))
        elif len(output_data) > n:
            output_data = output_data[:n]

        out_tensor = SovereignTensor(output_data, tensor.shape, tensor.spectral_meta)

        # TAURUS memory update
        self.taurus.store(out_tensor, importance=tensor.resonance)

        # Fuse with top-k memory context (residual connection)
        context = self.taurus.recall_top_k(4)
        if context:
            # Average context and add as residual
            ctx_sum = [0.0] * n
            for ctx_tensor in context:
                for i in range(min(len(ctx_tensor.data), n)):
                    ctx_sum[i] += ctx_tensor.data[i]
            ctx_weight = 0.1 / len(context)
            fused = [
                output_data[i] + ctx_sum[i] * ctx_weight for i in range(n)
            ]
            out_tensor = SovereignTensor(fused, tensor.shape, tensor.spectral_meta)

        return out_tensor

    def get_attention_analysis(self) -> Dict[str, Any]:
        """Return interpretability metrics from last forward pass.

        Returns:
            Dictionary with attention entropy, max attention, and sparsity
            metrics per head.
        """
        if not self._attention_maps:
            return {"n_heads": 0, "head_analyses": []}

        analyses = []
        for i, attn in enumerate(self._attention_maps):
            # Entropy
            entropy = -sum(
                w * math.log(w + 1e-10) for w in attn if w > 0
            )
            max_attn = max(attn) if attn else 0.0
            sparsity = sum(1 for w in attn if w < 0.01) / max(len(attn), 1)
            analyses.append(
                {
                    "head": i,
                    "attention_entropy": entropy,
                    "max_attention": max_attn,
                    "attention_sparsity": sparsity,
                }
            )

        return {"n_heads": len(self._attention_maps), "head_analyses": analyses}

    def reset_memory(self) -> None:
        """Clear TAURUS working memory."""
        self.taurus.working_memory.clear()

    def __repr__(self) -> str:
        return (
            f"SovereignNeuroCore(d_model={self.d_model}, n_heads={self.n_heads}, "
            f"memory={self.taurus.size()})"
        )
